### ISSUE TITLE
Can this crate be used for text formatting?

### ISSUE THREAD

**Issue by YetAnotherMinion** [Marwes/pretty.rs#47]
I would like to format error messages from a compiler. I looked at a compiler in Haskell and thought their `Doc` based formatter to layout text mixed with code snippets looked really nice. 

I am having trouble figuring out how to get words to pack into a line. When I use intersperse on a long slice of word Documents, then I see each word printed on a new line.

```rust
fn reflow<T: AsRef<str> >(s: T) -> Doc<'static, BoxDoc<'static, ()>> {                                                                                                                                                                           
    Doc::intersperse(s.as_ref().split_whitespace().map(|x| Doc::as_string(x)), Doc::space())                                                
}

fn reflow_works() {
    let doc = reflow("this is a longer sentence that will certainly exceed the 80 character limit of regular terminals");
    println!("{:?}", doc);
    // Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Text("this"), Space), Text("is")), Space), Text("a")), Space), Text("longer")), Space), Text("sentence")), Space), Text("that")), Space), Text("will")), Space), Text("certainly")), Space), Text("exceed")), Space), Text("the")), Space), Text("80")), Space), Text("character")), Space), Text("limit")), Space), Text("of")), Space), Text("regular")), Space), Text("terminals"))
    println!("{:?}", format!("{}", doc.pretty(80)));
   // "this\nis\na\nlonger\nsentence\nthat\nwill\ncertainly\nexceed\nthe\n80\ncharacter\nlimit\nof\nregular\nterminals"
}

```

Do I need to implement my own render if I want to pack as many appended `Doc::Text` elements as possible into one line?

**Comment by Marwes** [comment]
You need to call `group` on a document to have them appear on the same line.

```rust
doc.group().pretty(80)
```

Should work. Needing to do it on the top document is consistent but isn't that user friendly so I have been think that behaviour should change.

**Comment by YetAnotherMinion** [comment]
No luck.

```rust
fn reflow<T: AsRef<str> >(s: T) -> Doc<'static, BoxDoc<'static, ()>> {                                                                                                                                                                           
    Doc::intersperse(s.as_ref().split_whitespace().map(|x| Doc::as_string(x)), Doc::space())                                                
}

fn reflow_works() {
    let doc = reflow("this is a longer sentence that will certainly exceed the 80 character limit of regular terminals");
    println!("{:?}", doc);
    // Group(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Append(Text("this"), Space), Text("is")), Space), Text("a")), Space), Text("longer")), Space), Text("sentence")), Space), Text("that")), Space), Text("will")), Space), Text("certainly")), Space), Text("exceed")), Space), Text("the")), Space), Text("80")), Space), Text("character")), Space), Text("limit")), Space), Text("of")), Space), Text("regular")), Space), Text("terminals")))
    println!("{:?}", format!("{}", doc.pretty(80)));
    // "this\nis\na\nlonger\nsentence\nthat\nwill\ncertainly\nexceed\nthe\n80\ncharacter\nlimit\nof\nregular\nterminals"
}
````

I will trace the `fitting` loop on my toy problem to understand why it thinks the append will not fit

**Comment by YetAnotherMinion** [comment]
Update: I am reading the paper and implementing the recursive version on a branch because for me that approach is simpler to learn the algorithm. I will then attempt to transform the recursive impl into an iterative version to compare it with yours.

**Comment by Marwes** [comment]
Looking at https://hackage.haskell.org/package/prettyprinter it seems that it lays out multiple words on a single line as long as there is room which seems to be what you want. In this crate it will "give up" and put newlines as every separator if a "group" does not fit (unless a nested group fits that is).

It is probably better to try andfit as much as possible per line though if that is possible to integrate in this crate.

**Comment by YetAnotherMinion** [comment]
I got a naive (recursive) version working. There is the performance problem of flatten and union being eagerly evaluated.
https://gist.github.com/YetAnotherMinion/ac4c62e64ce58bef78cacd8771d6704d

My version is good enough for my compiler errors messages. I have decided to not deliver a patch to your library at this time because my use case is for a job. I do not receive any story points in JIRA for open source work. I would like to come back to this after crunch is over, but that will be at least two months.

Thank you for being quickly responsive.
